crates/sp-lightclient/src/lib.rs

```diff
@@ -79,22 +81,140 @@ pub trait Storage<Header: HeaderT> {
+        let best_header = self.best_header();
+        let current_finalized_number = self.finalized_head().0;
+
+        // ensure we have imported at least k-depth number of headers
+        let number_to_finalize = match best_header.header.number().checked_sub(&k_depth) {
+            // we have not progressed that far to finalize yet.
+            None => {
+                // explicitly set finalized head to genesis if it is not genesis.
+                // this could happen when the re-org happens and a heavier smaller chain becomes canonical and
+                // previous fork was long enough to have finalized some heads.
+                if current_finalized_number > Zero::zero() {
+                    self.finalize_header(self.heads_at_number(Zero::zero())[0]);
+                }
+                return Ok(());
+            }
+            Some(number) => number,
+        };
+
+        // we want to finalize the header from the current finalized header until the k-depth number of the best.
+        // 1. An ideal scenario, the current finalized head is one number less than number to be finalized but
+        // 2. if there was a re-org to longer chain when new header was imported, we do not want to miss
+        //    pruning fork headers between current and to be finalized number. So we go number by number and prune fork headers.
+        // 3. if there was a re-rg to a shorter chain when the new header was imported, then as per k-depth
+        //    we would have finalized more headers and, unfortunately, would have pruned any valid forks
+        //    There is not much we can do at this point then to just move the finalized head back and proceed.
+        match number_to_finalize.cmp(&current_finalized_number) {
+            // move the finalized head back to the expected head and return
+            Ordering::Less => {
+                let head_hash = self.heads_at_number(number_to_finalize)[0];
+                self.finalize_header(head_hash);
+            }
+            // nothing to do as we finalized the header already.
+            Ordering::Equal => (),
+            // move step by step from current to expected and prune all fork headers
+            Ordering::Greater => {
+                let mut current_finalized_number = current_finalized_number;
+
+                while current_finalized_number < number_to_finalize {
+                    current_finalized_number = current_finalized_number
+                        .checked_add(&One::one())
+                        .expect("should not overflow");
+
+                    // find the ancestor of the best header at k-depth
+                    let ancestor_header = self
+                        .ancestor_of_header_at_depth(current_finalized_number, &best_header)
+                        .expect("ancestor must exist at this point in time");
+
+                    // finalize the ancestor
+                    let mut number = *ancestor_header.header.number();
```

Comment:
This value is the same as `current_finalized_number` and would be easier to understand if written accordingly:
```suggestion
                    let mut finalized_number = current_finalized_number;
```